### PR TITLE
feat: support search-r1 training

### PR DESCRIPTION
**Pull request by LUMO666**
### Description
Support Search-r1 RL training on qwen2.5-3b. Search tool is based on local retriever server with wiki corpus. This PR supersedes #386  and uses a different source branch. Find previous discussion of the code in PR #386

### How has this been tested?
It has been test on qwen2.5-3b-instruct ,with 4 GPU.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -16,3 +16,4 @@ logs/*
 .venv
 uv.lock
 rlinf/envs/maniskill/assets
+examples/results
```
Comment (qurakchin) [inline .gitignore]: is this line needed?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,15 +0,0 @@
-repos:
```
Comment (qurakchin) [inline .pre-commit-config.yaml:1]: why delete this file?

File: examples/searchr1/config/qwen2.5-3b-tool-1node-eval.yaml [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+hydra:
+  run:
+    dir: .
+  output_subdir: null
+
+cluster:
+  num_nodes: 1
+  component_placement:
+    # actor,rollout,reward: all
```
Comment (qurakchin) [inline examples/searchr1/config/qwen2.5-3b-tool-1node-eval.yaml]: delete if not used.

File: examples/searchr1/config/qwen2.5-3b-tool-1node-eval.yaml [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+  gpu_memory_utilization: 0.8
+
+  model_dir: /path/to/eval/model
+  model_arch: qwen2.5
+  precision: ${actor.model.precision}      
+  enforce_eager: False         # if False, rollout engine will capture cuda graph, which will take more time to initialize.
+  distributed_executor_backend: mp   # ray or mp
+  disable_log_stats: False
+  detokenize: True            # Whether to detokenize the output. During RL we actually don't need to detokenize it. Can be set to True for debugging.
```
Comment (qurakchin) [inline examples/searchr1/config/qwen2.5-3b-tool-1node-eval.yaml:110]: add a comment to emphasize that if sampling_params.stop used, detokenize must be True.

File: examples/searchr1/config/qwen2.5-3b-tool-1node.yaml [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+  # mbs to do log prob inference, can be set to
+  # lower than rollout_batch_size_per_gpu to reduce
+  # memory usage
+  logprob_forward_micro_batch_size: 1 # ${.rollout_batch_size_per_gpu}
+
+  # val rollout mbs
+  val_rollout_batch_size_per_gpu: 1 # ${.rollout_batch_size_per_gpu}
+
+  recompute_logprobs: True
```
Comment (qurakchin) [inline examples/searchr1/config/qwen2.5-3b-tool-1node.yaml:56]: add a comment to emphasize that search r1 re-tokenize the llm output, so recompute_logprobs must be True.

File: rlinf/workers/rollout/sglang/sglang_worker.py [hunk 6]
```diff
@@ -68,6 +68,7 @@ def __init__(self, config: DictConfig, placement: ModelParallelComponentPlacemen
         self._collect_meta_stats = getattr(
             self._cfg.rollout, "collect_meta_stats", False
         )
+        self.is_eval = getattr(self._cfg.rollout, "is_eval", 0) == 1
```
Comment (qurakchin) [inline rlinf/workers/rollout/sglang/sglang_worker.py]: change to bool, true or false

File: examples/searchr1/download.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+# Copyright 2024 Bytedance Ltd. and/or its affiliates
```
Comment (qurakchin) [inline examples/searchr1/download.py]: is this copyright style right? @andylin-hao

File: examples/searchr1/config/qwen2.5-3b-tool-1node-eval.yaml (hunk 4 above)
Comment (qurakchin) [reply examples/searchr1/config/qwen2.5-3b-tool-1node-eval.yaml:110]: please add emphasized text to the rst doc

File: examples/searchr1/config/qwen2.5-3b-tool-1node.yaml (hunk 5 above)
Comment (qurakchin) [reply examples/searchr1/config/qwen2.5-3b-tool-1node.yaml:56]: please add emphasized text to the rst doc

File: examples/searchr1/config/qwen2.5-3b-tool-1node-eval.yaml [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+
+  # mbs to do log prob inference, can be set to
+  # lower than rollout_batch_size_per_gpu to reduce
+  # memory usage
+  logprob_forward_micro_batch_size: 1 # ${.rollout_batch_size_per_gpu}
+
+  # val rollout mbs
+  val_rollout_batch_size_per_gpu: 4 # ${.rollout_batch_size_per_gpu}
+
+  recompute_logprobs: True
```
Comment (qurakchin) [inline examples/searchr1/config/qwen2.5-3b-tool-1node-eval.yaml:52]: recompute_logprobs, grpo params, inference, actor are not needed in eval yaml.

**Review by qurakchin — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 9]
```diff
@@ -16,4 +16,4 @@ outputs/*
 *.so
 .venv
 uv.lock
-rlinf/envs/maniskill/assets
+rlinf/envs/maniskill/assets
```
Comment (qurakchin) [inline .gitignore:19]: What's the difference of this line? Do you change the end line?

File: .pre-commit-config.yaml [hunk 10]
```diff
@@ -11,5 +11,4 @@ repos:
     hooks: # support hooks
     -   id: check-message  # requires commit-msg hook
     -   id: check-branch
-    -   id: check-commit-signoff
-
+    -   id: check-commit-signoff
```
Comment (qurakchin) [inline .pre-commit-config.yaml:14]: same as the last comment

File: README.md [hunk 11]
```diff
@@ -30,8 +30,9 @@ RLinf is a flexible and scalable open-source infrastructure designed for post-tr
 
 
 ## What's NEW!
-- [2025/11] 🔥 RLinf supports reinforcement learning fine-tuning for [CALVIN](https://github.com/mees/calvin). Doc: [RL on CALVIN](https://rlinf.readthedocs.io/en/latest/rst_source/examples/calvin.html).
-- [2025/11] 🔥 RLinf supports reinforcement learning fine-tuning for [IsaacLab](https://github.com/isaac-sim/IsaacLab). Doc: [RL on IsaacLab](https://rlinf.readthedocs.io/en/latest/rst_source/examples/isaaclab.html). 
+- [2025/12] 🔥 RLinf supports agentic reinforcement learning on [Search-R1](https://github.com/PeterGriffinJin/Search-R1). Doc: [Search-R1](https://rlinf.readthedocs.io/en/latest/rst_source/examples/searchr1.html)
```
Comment (qurakchin) [inline README.md]: The dot sigh in the end of the line is missing.

File: rlinf/config.py [hunk 12]
```diff
@@ -889,52 +908,45 @@ def validate_cfg(cfg: DictConfig) -> DictConfig:
...
         cfg = validate_embodied_cfg(cfg)
+    elif cfg.runner.task_type == "reasoning_eval":
+        cfg = validate_reasoning_eval_cfg(cfg)
     elif cfg.runner.task_type == "reasoning":
         cfg = validate_reasoning_cfg(cfg)
     elif cfg.runner.task_type == "coding_online_rl":
         cfg = validate_coding_online_rl_cfg(cfg)
 
-    if cfg.algorithm.adv_type in ("grpo", "reinpp_baseline"):
+    if cfg.algorithm.adv_type in ("grpo", "reinpp_baseline") and not cfg.rollout.get(
```
Comment (qurakchin) [inline rlinf/config.py:897]: use cfg.runner.task_type == "reasoning_eval" instead of cfg.rollout.is_eval. the cfg.runner.task_type is formal to indicate no training in this cfg.

File: rlinf/config.py [hunk 13]
```diff
@@ -889,52 +908,45 @@ def validate_cfg(cfg: DictConfig) -> DictConfig:
     )
     if cfg.runner.task_type == "embodied":
         cfg = validate_embodied_cfg(cfg)
+    elif cfg.runner.task_type == "reasoning_eval":
+        cfg = validate_reasoning_eval_cfg(cfg)
```
Comment (qurakchin) [inline rlinf/config.py]: is all the change behind for validate the training? if true, we can return there.

File: .gitignore (hunk 9 above)
Comment (LUMO666) [reply .gitignore:19]: delete 1 unnecessary line at last commit

File: README.md [hunk 14]
```diff
@@ -30,8 +30,9 @@ RLinf is a flexible and scalable open-source infrastructure designed for post-tr
 
 
 ## What's NEW!
-- [2025/11] 🔥 RLinf supports reinforcement learning fine-tuning for [CALVIN](https://github.com/mees/calvin). Doc: [RL on CALVIN](https://rlinf.readthedocs.io/en/latest/rst_source/examples/calvin.html).
```
Comment (andylin-hao) [inline README.md:35]: There is a merge issue here. You have deleted CALVIN and add a duplicate BEHAVIOR below

File: rlinf/hybrid_engines/sglang/common/sgl_scheduler.py [hunk 15]
```diff
@@ -188,10 +257,54 @@ def init_rlinf_worker(
...
+            self.weight_norm_dict = weight_norm_dict
+
+            # avoid release memory before norm kernel launch (gpu is async from cpu)
+            torch.cuda.synchronize()
+
         self._rlinf_worker.log_info(
             f"Running Scheduler dp rank {self._rlinf_worker.get_parent_rank()}, tp rank {self.tp_rank}, corresponding actor weight rank = {self.actor_weight_rank}"
         )
 
+    def init_rlinf_worker_no_sync(
```
Comment (Lin-xs) [inline rlinf/hybrid_engines/sglang/common/sgl_scheduler.py]: Add an `if` statement in the `init_rlinf_worker`, there is no need to add a new initialization method.

File: rlinf/hybrid_engines/sglang/common/sgl_scheduler.py [hunk 16]
```diff
@@ -93,9 +94,47 @@ def cuda_info(self, text: str = ""):
...
+    def resume_memory_occupation(self, recv_req: ResumeMemoryOccupationReqInput):
+        assert self.is_weight_offloaded is True, "Weight has been onloaded!"
+        self.is_weight_offloaded = False
+        self._rlinf_worker.log_info("Running Scheduler onload")
+        result = super().resume_memory_occupation(recv_req)
+        if not self.is_sync_weight:
+            model = self.tp_worker.worker.model_runner.model
+            model.load_state_dict(self.cpu_state_dict)
+
+            if self.weight_norm_dict is not None:
```
Comment (Lin-xs) [inline rlinf/hybrid_engines/sglang/common/sgl_scheduler.py]: Write the parameter checking process as a function.

File: rlinf/workers/rollout/sglang/sglang_worker.py [hunk 17]
```diff
@@ -240,13 +253,25 @@ async def init_worker(self):
...
+        if self._cfg_rollout.validate_weight:
             await self._validate_weight_at_first()
         if self._placement.is_collocated:
             await self.offload_engine()
         if self._use_auto_scheduler:
             asyncio.create_task(self._scheduler.main_loop())
 
+    async def init_worker_no_sync(self):
+        self._init_engine()
+        await self._engine.tokenizer_manager.run_task_method(
```
Comment (andylin-hao) [inline rlinf/workers/rollout/sglang/sglang_worker.py]: I don't see a fundamental difference between this and the original `init_worker`. Can they simply be one function with an argument `no_sync`? The same goes for `init_rlinf_worker_no_sync` in the scheduler.

File: rlinf/workers/rollout/sglang/sglang_worker.py [hunk 18]
```diff
@@ -44,17 +47,28 @@
 
 
 class SGLangWorker(Worker):
-    def __init__(self, config: DictConfig, placement: ModelParallelComponentPlacement):
+    def __init__(
+        self,
+        config: DictConfig,
+        placement: ModelParallelComponentPlacement,
+        config_rollout: DictConfig = None,
```
Comment (guozhen1997) [inline rlinf/workers/rollout/sglang/sglang_worker.py]: why add `config_rollout` and `sampling_params` instead of using `config.rollout` and `config.algorithm.sampling_params`
